Improve documentation (#171)
aiidateam/aiida-common-workflows: docs/source/workflows/base/relax.rst

Common relax workflow
---------------------

The common relax workflow allows to perform structural relaxation towards the most energetically favorable configuration through a
common interface shared by eleven quantum engines: Abinit, BigDFT, CASTEP, CP2K, FLEUR, Gaussian, NWChem, ORCA, Quantum ESPRESSO,
Siesta, VASP.
Relaxation of both molecules and crystals is supported, with the caveat that ORCA and Gaussian can not perform relaxation

> on extended systems.

Reviewer: nuance that Gaussian can, but the plugin doesn't suppor ti

> The common relax workflow allows to perform structural relaxation towards the most energetically favorable configuration through a
> common interface shared by eleven quantum engines: Abinit, BigDFT, CASTEP, CP2K, FLEUR, Gaussian, NWChem, ORCA, Quantum ESPRESSO,
> Siesta, VASP.
> Relaxation of both molecules and crystals is supported, with the caveat that ORCA and Gaussian can not perform relaxation
> on extended systems.

Reviewer: ```suggestion
The common relax workflow performs a geometric optimization of a molecule or extended system towards the most energetically favorable configuration.
It defines a common interface that is currently implemented by eleven quantum engines: Abinit, BigDFT, CASTEP, CP2K, FLEUR, Gaussian, NWChem, ORCA, Quantum ESPRESSO, Siesta, VASP.
The ORCA and Gaussian implementations only support the optimization of molecules.
```

In the initial page of this documentation the general instructions for the :ref:`submission of a common workflow <how-to-submit>`

> are presented, covering both the use of the built-in command line interface (CLI) and the creation of a submission scripts.

Reviewer: ```suggestion
are presented, covering both the use of the built-in command line interface (CLI) and custom submission scripts.
```

> Since the CLI does not expose the full functionalities of the interface, the documentation of this section is focused

Reviewer: ```suggestion
Since the CLI does not expose the full functionalities of the interface, this section focuses
```

> on the creation of input scripts and explains how to have full control on the relaxation process through the common interface.

Reviewer: ```suggestion
on the creation of submission scripts and explains how to have full control on the relaxation process through the common interface.
```

> A mention on the CLI functionalities is reported at the end of the section.

Reviewer: ```suggestion
The CLI is explained at the end of the section.
```

Relaxation inputs
.................

> A typical submission script of the common relax workflow is:

Reviewer: ```suggestion
A typical script for the submission of common relax workflow could look something like the following:
```

.. code:: python

    from aiida.engine import submit
    from aiida.plugin import WorkflowFactory

    RelaxWorkChain = WorkflowFactory('common_workflows.relax.<implementation>')  # Load the relax workflow implementation of choice.

    # DEFINITION OF <RELAXATION INPUTS>

    input_generator = RelaxWorkChain.get_input_generator()
    builder = input_generator.get_builder(<RELAXATION INPUTS>)
    submit(builder)

> Please note that the quantum engine to be used for the relaxation is selected by the call to the ``WorkflowFactory``. One
> ``<implementation>`` is available for each supported quantum engine (for instance use
> ``quantum_espresso`` to select Quantum ESPRESSO). The list can be explored via the command
> ``verdi plugin list aiida.workflows`` and looking for entry point that starts with ``common_workflows.relax``.

Reviewer: ```suggestion
The quantum engine that will be used for the relaxation is determined by the entry point that is passed to the ``WorkflowFactory``.
To select a particular quantum engine, you should replace ``<implementation>`` in the string ``'common_workflows.relax.<implementation>'`` with the corresponding entry point name of the quantum engine.
To see a list of all available entry points, call ``verdi plugin list aiida.workflows``.
Any entry point that starts with ``common_workflows.relax.`` can be used, for example ``common_workflows.relax.quantum_espresso`` corresponds to the implementation using Quantum ESPRESSO.
```

> In the following, it is reported the complete list of ``<RELAXATION INPUTS>``
> (accepted arguments of ``RelaxWorkChain.get_input_generator().get_builder``).
> Only the first two in the list (``structure`` and ``engines``) are positional arguments, the rest must be invoked using the
> corresponding keyword.

Reviewer: ```suggestion
The ``RelaxWorkChain.get_input_generator().get_builder`` method takes a number of required and optional arguments (referenced as ``<RELAXATION INPUTS>`` in the example submission script), which are listed below.
Only ``structure`` and ``engines`` can be specified as a positional argument, all others have to be defined as a keyword argument.
```

* ``structure``. (Type: an AiiDA `StructureData`_ instance,
  the common data format to specify crystal structures and molecules in AiiDA).
  The structure to relax.

* ``engines``. (Type: a Python dictionary).
  It specifies the codes and the corresponding computational resources for each step of the relaxation
  process. A typical example is:

  .. code:: python

        engines = {
              'relax': {

>                   'code': <CODE>,  # An identifier of a `Code` configured for the selected quantum engine plugin

Reviewer: ```suggestion
                  'code': <CODE>,  # An identifier (PK, UUID or full label) of a `Code` configured for the selected quantum engine plugin, e.g., ``'pw-v6.5@localhost'`.
```

                  'options': {
                      'resources': {
                          'num_machines': 1,  # Number of machines/nodes to use
                      },
                      'max_wallclock_seconds': 3600,  # Number of wallclock seconds to request from the scheduler for each job
                  }
              }
          }

  Typically one single executable is sufficient to perform the relaxation. However, there are cases in which
  two or more codes in the same simulation package are required to achieve the final goal, as for example in
  the case of FLEUR. In order to explore the steps of the relaxation process, the ``input_generator``
  provides an inspection method:

  .. code:: python

        input_generator.get_engine_types()

>   For each ``engine_type`` (for instance "inpgen" of FLEUR),

Reviewer: ```suggestion
  For each ``engine_type`` (for instance ``inpgen`` for FLEUR),
```

  details on the required code and resources can be obtained with:

  .. code:: python

        input_generator.get_engine_type_schema("inpgen")

* ``protocol``. (Type: a Python string).
  A single string summarizing the computational accuracy of the underlying DFT calculation and relaxation algorithm.

>   Three protocol names are defined and implemented for each code: ‘fast’, ‘moderate’ and ‘precise’.
>   The details of how each implementation translates a protocol string into a choice of parameters is code dependent, or more
>   specifically, they depend on the implementation choices of the corresponding AiiDA plugin.
>   However the chosen parameters respect the meaning of the corresponding string: a possibly unconverged (but still meaningful)
>   run that executes rapidly for testing is obtained with the ‘fast’ protocol; the ‘moderate’ protocol is
>   a safe choice for prototyping and preliminary studies; and a set of converged parameters that might result in an
>   expensive simulation but provides converged results is obtained with the ‘precise’ protocol.

Reviewer: ```suggestion
  Three protocol names are defined and implemented for each code: ``fast``, ``moderate`` and ``precise``.
  The details of how each implementation translates a protocol string into a choice of parameters is code dependent, or more
  specifically, they depend on the implementation choices of the corresponding AiiDA plugin.
  However the chosen parameters respect the meaning of the corresponding string:
    * ``fast``: a possibly unconverged (but still meaningful) run that executes rapidly for testing.
    * ``moderate``: a safe choice for prototyping and preliminary studies.
    * ``precise``: uses parameters that might result in an computationally expensive simulation but provides well converged results.
```

  (Type: AiiDA ``StructureData``).

* ``total_energy``. The total energy in eV associated to the relaxed structure
  (or initial structure in case no relaxation is performed).
  In general, even for calculations performed with the same code, there is no guarantee to have comparable
  energies in different runs if the numerical parameters determined by the input generator change
  (because, for instance, structures with different volumes are passed). However, in combination with the
  input argument ``reference_workchain``, energies from different relaxation
  runs become comparable, and their energy difference is well defined. (Type: AiiDA ``Float``).

* ``stress``.   The final stress tensor in eV/Å^3.
  Returned only when a variable-cell relaxation is performed.
  (Type: AiiDA ``Float``).

* ``total_magnetization``. The total magnetization in
  μB (Bohr-magneton) units.  Returned only for magnetic calculations.
  (Type: AiiDA ``Float``).

CLI options
...........

The use of the CLI for the submission of a common workflow is reported in the :ref:`main page <how-to-submit>` of this documentation.

> For the relaxation workflow:

Reviewer: Okay to hard-code this for now, but from experience this is a pain to maintain. Everytime the CLI is changed, this has to be probably updated and that is easy to forget. For `aiida-core` I implemented a pre-commit hook that will automatically update this. Just mentioning this so we can add an issue to add this at some point. Especially once we start documenting more CLI commands.
